runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkJobServerDriver.java

```diff
@@ -107,12 +107,19 @@ public static FlinkJobServerDriver fromConfig(ServerConfiguration configuration)
         new ThreadFactoryBuilder().setNameFormat("flink-runner-job-server").setDaemon(true).build();
     ListeningExecutorService executor =
         MoreExecutors.listeningDecorator(Executors.newCachedThreadPool(threadFactory));
-    ServerFactory jobServerFactory = ServerFactory.createWithPortSupplier(() -> configuration.port);
+    ServerFactory jobServerFactory =
+        ServerFactory.createWithPortSupplier(() -> getPort(configuration.host, configuration.port));
     ServerFactory artifactServerFactory =
-        ServerFactory.createWithPortSupplier(() -> configuration.artifactPort);
+        ServerFactory.createWithPortSupplier(
+            () -> getPort(configuration.host, configuration.artifactPort));
     return create(configuration, executor, jobServerFactory, artifactServerFactory);
   }
 
+  private static int getPort(String host, int port) {
+    // If host is empty then use dynamic port
+    return Strings.isNullOrEmpty(host) ? 0 : port;
```

Comment:
Why do we not use the provided port when the host is null? I think it would make sense that the user doesn't care about the host name but wants to bind to a specific port.

Reply:
We need to be able to use dynamic port to create server in case of test.
Port supplier is used when using allocateAndCreate which is expected to use a dynamic port. As port was fixed in portSupplier, this functionality was broken.
I am using host as the marker to create dynamic port or static as we do in [L181 ](https://github.com/apache/beam/pull/6341/files#diff-e52a25378b181af8cecbf3d274946fd0L181)

Reply:
Imagine the user started the JobServer with `--job-port 1234`, then it would bind to a different port due to `--job-host` being empty. We should simply pass in `0` as the port in tests. This would then bind to an arbitrary port. Additionally, we should clarify the semantics through `ServerFactoryTest` (can be a follow-up).

Reply:
Here is the PR #6469.